lib/Interpreter/IncrementalCUDADeviceCompiler.h

```diff
@@ -35,120 +36,78 @@ namespace llvm {
-    /// cuda NVIDIA fatbinary are found.
+    ///\brief Is true if the second interpreter instance was created and the
+    /// NVIDIA fatbin tool was found.
     bool m_Init = false;
 
-    ///\brief Path to the folder, where all files will put in. Ordinary the tmp
-    /// folder. Have to end with a separator. Can be empty.
-    const std::string m_FilePath;
-    ///\brief Path to the fatbin file, which will used by the CUDACodeGen.
-    const std::string m_FatbinFilePath;
-    ///\brief Path to a empty dummy.cu file. The file is necessary to generate
-    /// PTX code from the pch files.
-    const std::string m_DummyCUPath;
-    ///\brief Path to the PTX file. Will be reused for every PTX generation.
-    const std::string m_PTXFilePath;
-    ///\brief Will be used to generate .cu and .cu.pch files.
-    const std::string m_GenericFileName;
-
-    ///\brief Path to the clang++ compiler, which will used to compile the pch
-    /// files and the PTX code. Should be in same folder, as the cling.
-    std::string m_ClangPath;
-    ///\brief Path to the NIVDIA tool fatbinary.
-    std::string m_FatbinaryPath;
-
-    ///\brief Contains information about all include paths.
-    ///
-    std::shared_ptr<clang::HeaderSearchOptions> m_HeaderSearchOptions;
-
-    ///\brief get copy of m_Counter
-    ///
-    ///\returns copy of m_Counter
-    unsigned int getCounterCopy() { return m_Counter; }
+    ///\brief Contains the PTX code of the current input
+    llvm::SmallString<1024> m_PTX_code;
 
-    ///\brief Generate the dummy.cu file and set the paths of m_PTXFilePath and
-    /// m_GenericFileName.
-    ///
-    ///\returns True, if it created a dummy.cu file.
-    bool generateHelperFiles();
-
-    ///\brief Find the path of the clang and the NIVDIA tool fatbinary. Clang
-    /// have to be in the same folder as cling.
-    ///
-    ///\param [in] invocationOptions - Can contains a custom path to the cuda
-    ///       toolkit
-    ///
-    ///\returns True, whether clang and fatbinary was found.
-    bool findToolchain(const cling::InvocationOptions& invocationOptions);
+    ///\brief pointer to the buffer in which the fatbinary code is to be stored
+    std::shared_ptr<llvm::SmallVectorImpl<char>> m_Fatbin_code;
 
     ///\brief Add the include paths from the interpreter runtime to a argument
     /// list.
     ///
     ///\param [in,out] argv - The include commands will append to the argv
     /// vector.
-    void addHeaderSearchPathFlags(llvm::SmallVectorImpl<std::string>& argv);
-
-    ///\brief Start an clang compiler with nvptx backend. Read the content of
-    /// cling.cu and compile it to a new PCH file. If predecessor PCH file is
-    /// existing, it will included.
+    ///\param [in] headerSearchOptions - Contains information about all include
+    /// paths.
+    void addHeaderSearchPathFlags(
+        std::vector<std::string>& argv,
+        const std::shared_ptr<clang::HeaderSearchOptions> headerSearchOptions);
```

Comment:
I think taking a `const std::shared_ptr<clang::HeaderSearchOptions> &headerSearchOptions` might make more sense - doesn't incr/decr the use count just for arg passing.